### PR TITLE
feat: add sqs-batch middleware

### PR DESCRIPTION
**Pull request by brettstack**
Related issue: #467

TODO:
- [x] Resolve dangling underscore and parameterized function overrides discussion

### CODE REVIEW

**Comment by willfarrell** [comment]
Looks good.

Are there any required settings that need to be set on the SQS for this to work as expected (ie `BatchSize`)? If so we should document.

**Comment by brettstack** [comment]
Excellent question. Batch size is 10 by default for SQS.

**Comment by brettstack** [comment]
Tests are failing in node 10 due to `Promise.allSettled` not being available. How do we skip for this version?

**Comment by willfarrell** [comment]
We can’t skip that version till AWS does. Please add in a poly fill for now.

**Comment by willfarrell** [comment]
Thanks for the great work. 

@lmammino  `sqs` or `sqs-batch`?

**Comment by brettstack** [comment]
Added the polyfill to the tests. To be clear: the middleware itself doesn't use `allSettled`, but the user needs to return a `Promise.allSettled()` value from their Lambda handler. So anyone wanting to use this middleware in Node < 12 will need to add a polyfill. I'll mention this in the docs.

**Comment by brettstack** [comment]
`sqs-partial-batch-failure` would be the most descriptive. Or even:

`sqs-partial-batch-failure-cleaner`
`sqs-partial-batch-failure-handler`
`sqs-graceful-partial-batch-failure`

File: packages/sqs-batch/package.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    "aws-sdk": ">=2.221.1"
+  },
+  "devDependencies": {
+    "@middy/core": "^1.0.0-beta",
+    "@serverless/event-mocks": "^1.1.1",
+    "promise.allsettled": "^1.0.2"
+  },
+  "gitHead": "7a6c0fbb8ab71d6a2171e678697de9f237568431",
+  "dependencies": {
+    "jest": "^25.1.0"
```
Comment (vladgolubev) [inline packages/sqs-batch/package.json]: Is this supposed to be a dependency?

**Comment by willfarrell** [comment]
@lmammino Was just thinking. Should we add a section to the middleware readmes that lets people know who to flag when they have a question or find an issue with it?

File: packages/sqs-batch/package.json (hunk 1 above)
Comment (willfarrell) [reply packages/sqs-batch/package.json]: `devDependencies` makes more sense to me, I'm guessing it was copied from somewhere.

Comment (brettstack) [reply packages/sqs-batch/package.json]: Absolutely not. Good catch.

File: packages/sqs-batch/package.json [hunk 2]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@middy/sqs-batch",
+  "version": "1.0.0-beta",
+  "description": "SQs batch error handler middleware for the middy framework",
```
Comment (vladgolubev) [inline packages/sqs-batch/package.json]:
```suggestion
  "description": "SQS batch error handler middleware for the middy framework",
```

Comment (brettstack) [reply packages/sqs-batch/package.json]: Ugh. The shift key on my shitty MacBook butterfly keyboard is going out. I'll fix this evening.

**Comment by lmammino** [comment]
Sorry for the delay on this! Busy days... I should have some time this afternoon to have a proper look and get back to you with some feedback!

**Comment by lmammino** [comment]
> Thanks for the great work.
>
> @lmammino sqs or sqs-batch?

I like `sqs-batch` between the 2. But between those ones:

> Some other options:
>
> sqs-partial-batch-failure-handler ❌handler is ambiguous
> sqs-partial-batch-failure-cleaner ✅cleaner is descriptive
> sqs-graceful-partial-batch-failure 👈I like graceful
> sqs-partial-batch-failure 🩳shortest

I like `sqs-partial-batch-failure` :)

> @lmammino Was just thinking. Should we add a section to the middleware readmes that lets people know who to flag when they have a question or find an issue with it?

We could, let's maybe create a dedicate issue / PR for this. What do you think?

File: packages/sqs-batch/README.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+
+## Sample usage
+
+```javascript
+const middy = require('@middy/core')
+const sqsBatch = require('@middy/sqs-batch')
+
+const originalHandler = (event, context, cb) => {
+  const recordPromises = Records.map(async (record, index) => { /* Custom message processing logic */ })
```
[the remark this answers is not in the document]
Comment (brettstack) [reply packages/sqs-batch/README.md]: Good catch. I meant to have `event.Records` 🙃

File: packages/sqs-batch/index.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
+/* eslint no-underscore-dangle:0 */
```
[the remark this answers is not in the document]
Comment (brettstack) [reply packages/sqs-batch/index.js]: The non-dangled version is defined as a parameter on the exported function to allow customers to override default functionality. I can remove this if you prefer, or rename those with some other prefix or suffix.

[the remark this answers is not in the document]
Comment (lmammino) [reply packages/sqs-batch/index.js]: Thanks for being available on changing this. I think in other middleware we just call that `defaultX`. Maybe this kind of prefix would be a bit more explicit. What do you think?

[the remark this answers is not in the document]
Comment (brettstack) [reply packages/sqs-batch/index.js]: Done!

**Comment by brettstack** [comment]
I think we're good to go

**Comment by lmammino** [comment]
Fantastic, thanks for your patience @brettstack! I should be able to have a release in the next few hours!

**Pull request merged**